core/services/ocr2/plugins/threshold/plugin.go

```diff
@@ -0,0 +1,79 @@
+package threshold
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+	libocr2 "github.com/smartcontractkit/libocr/offchainreporting2plus"
+
+	decryptionPlugin "github.com/smartcontractkit/tdh2/go/ocr2/decryptionplugin"
+	decryptionPluginConfig "github.com/smartcontractkit/tdh2/go/ocr2/decryptionplugin/config"
+	"github.com/smartcontractkit/tdh2/go/tdh2/tdh2easy"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type ThresholdServicesConfig struct {
+	DecryptionQueue    decryptionPlugin.DecryptionQueuingService
+	KeyshareWithPubKey []byte
+	ConfigParser       decryptionPluginConfig.ConfigParser
+}
+
+func NewThresholdService(sharedOracleArgs *libocr2.OCR2OracleArgs, conf *ThresholdServicesConfig) (job.ServiceCtx, error) {
+	publicKey, privKeyShare, err := UnmarshalKeys(conf.KeyshareWithPubKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal threshold key share with public key")
+	}
+
+	// The key generation tooling ensures that key IDs correspond to the oracle's index,
+	// therefore an identity mapping is used when creating the threshold reporting plugin.
+	// maxNumNodes is selected such that it will always be larger than the number of nodes in the DON.
+	oracleToKeyShare := make(map[commontypes.OracleID]int)
+	maxNumNodes := 100
+	for i := 0; i <= maxNumNodes; i++ {
+		oracleToKeyShare[commontypes.OracleID(i)] = i
+	}
+
+	sharedOracleArgs.ReportingPluginFactory = decryptionPlugin.DecryptionReportingPluginFactory{
+		DecryptionQueue:  conf.DecryptionQueue,
+		ConfigParser:     conf.ConfigParser,
+		PublicKey:        &publicKey,
+		PrivKeyShare:     &privKeyShare,
+		OracleToKeyShare: oracleToKeyShare,
+		Logger:           sharedOracleArgs.Logger,
+	}
+
+	thresholdReportingPluginOracle, err := libocr2.NewOracle(*sharedOracleArgs)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to call NewOracle to create a Threshold Reporting Plugin")
+	}
+
+	return job.NewServiceAdapter(thresholdReportingPluginOracle), nil
+}
+
+type KeyshareWithPubKey struct {
+	PublicKey       json.RawMessage //tdh2easy.PublicKey
```

Comment:
I still don't get why you can't just put those types here. Unmarshal literally does what you're doing below for all fields. Can you explain that again, please?

Reply:
There is additional logic required to unmarshal the JSON into the tdh2easy.PublicKey & tdh2easy.PrivateKey types beyond a simple JSON unmarshal.

Reply:
Yes but that logic sits in PublicKey.Unmarshal() and that should be called automatically for all fields if you unmarshal the struct. Can you confirm that it works (or doesn't)?

Reply:
After a quick search, it appears this is mostly correct.  json.Unmarshal will call the UnmarshalJSON method for the types contained within the struct.

Unfortunately, it appears tdh2easy uses the method name `Unmarshal` instead of `UnmarshalJSON`, so that will not work in this case. We need to get a PR merged to tdh2easy to change the method name `Unmarshal` to `UnmarshalJSON`.

Is is possible that we could do this in the final integration PR to avoid blocking this PR on the necessary change to tdh2?

Reply:
sure, np